skyportal/handlers/api/public_pages/public_source_page.py

```diff
@@ -0,0 +1,267 @@
+from ...base import BaseHandler
+
+from ....models import PublicSourcePage, Group, Stream, Classification, Photometry
+from ....utils.thumbnail import get_thumbnail_alt_link, get_thumbnail_header
+
+log = make_log('api/public_source_page')
+
+
+def calculate_hash(data):
+    return joblib.hash(json.dumps(data, sort_keys=True))
+
+
+def process_thumbnails(thumbnails, ra, dec):
+    for index, thumbnail in enumerate(thumbnails):
+        alt, link = get_thumbnail_alt_link(thumbnail["type"], ra, dec)
+        header = get_thumbnail_header(thumbnail["type"])
+        thumbnails[index] = {
+            "type": thumbnail["type"],
+            "public_url": thumbnail["public_url"],
+            "alt": alt,
+            "link": link,
+            "header": header,
+        }
+    return thumbnails
+
+
+class PublicSourcePageHandler(BaseHandler):
+    @permissions(['Manage sources'])
+    async def post(self, source_id):
+        """
+        ---
+          description:
+            Create a public page for a source, with given options,
+            only if this page does not already exist
+          tags:
+            - public_source_page
+          parameters:
+            - in: path
+              name: source_id
+              schema:
+                type: string
+                required: true
+                description: The ID of the source from which to create a public page
+          requestBody:
+            content:
+                application/json:
+                    schema:
+                        type: object
+                        properties:
+                            options:
+                                type: object
+                                required: true
+                                description: Options to manage data to display publicly
+          responses:
+            200:
+              content:
+                application/json:
+                  schema: Success
+            400:
+              content:
+                application/json:
+                  schema: Error
+        """
+        data = self.get_json()
+        if data is None or data == {}:
+            return self.error("No data provided")
+        if source_id is None:
+            return self.error("Source ID is required")
+        options = data.get("options")
+        if options is None:
+            return self.error("Options are required")
+
+        with self.Session() as session:
+            group_ids = options.get("groups")
+            stream_ids = options.get("streams")
+
+            # get source
+            source = await get_source(
+                source_id,
+                self.associated_user_object.id,
+                session=session,
+                include_thumbnails=True,
+            )
+            if source is None:
+                return self.error("Source not found", status=404)
+            data_to_publish = {
+                "ra": round(source["ra"], 6) if source["gal_lon"] else None,
```

Comment:
Just checking: how has the rounding of these values been decided? Based on what is on the frontend?

Reply:
Yes exactly i follow the private source display round.